aiida_lammps/workflows/base.py

```diff
@@ -0,0 +1,351 @@
+from aiida.common import AttributeDict, InputValidationError
+from aiida.engine import (
+    BaseRestartWorkChain,
+    ProcessHandlerReport,
+    process_handler,
+    while_,
+)
+import numpy as np
+
+from aiida_lammps.calculations.base import LammpsBaseCalculation
+from aiida_lammps.utils import RestartTypes
+
+
+class LammpsBaseWorkChain(BaseRestartWorkChain):
+    """Base workchain for calculations using LAMMPS"""
+
+    _process_class = LammpsBaseCalculation
+
+    @classmethod
+    def define(cls, spec):
+        """Define the process specification."""
+        # yapf: disable
+        super().define(spec)
+        spec.expose_inputs(LammpsBaseCalculation, namespace='lammps')
+        spec.input(
+            "store_restart",
+            valid_type=orm.Bool,
+            default=lambda: orm.Bool(False),
+            required=False,
+            help="""
+            Whether to store the restartfile in the repository.
+            """,
+        )
+        spec.expose_outputs(LammpsBaseCalculation)
+        spec.outline(
+            cls.setup,
+            cls.validate_parameters,
+            while_(cls.should_run_process)(
+                cls.run_process,
+                cls.inspect_process,
+            ),
+            cls.results,
+        )
+
+        spec.exit_code(
+            300,
+            "ERROR_UNRECOVERABLE_FAILURE",
+            message="""
+            The calculation failed with an unidentified unrecoverable error.
+            """,
+        )
+        # yapf: enable
+
+    def setup(self):
+        """Call the ``setup`` of the ``BaseRestartWorkChain`` and create the inputs dictionary in ``self.ctx.inputs``.
+
+        This ``self.ctx.inputs`` dictionary will be used by the ``BaseRestartWorkChain`` to submit the calculations
+        in the internal loop.
+
+        The ``parameters`` and ``settings`` input ``Dict`` nodes are converted into a regular dictionary and the
+        default namelists for the ``parameters`` are set to empty dictionaries if not specified.
+        """
+        super().setup()
+        self.ctx.inputs = AttributeDict(
+            self.exposed_inputs(LammpsBaseCalculation, "lammps")
+        )
+        self.ctx.inputs.settings = (
+            self.ctx.inputs.settings.get_dict() if "settings" in self.ctx.inputs else {}
+        )
+
+    def validate_parameters(self):
+        """Validate the parameters for the workchain"""
+
+        if "store_restart" in self.ctx.inputs:
+            self.ctx.inputs.settings.store_restart = self.ctx.inputs.store_restart.value
+
+        if "script" not in self.ctx.inputs and any(
+            key not in self.ctx.inputs
+            for key in ["parameters", "potential", "structure"]
+        ):
+            raise InputValidationError(
+                "Unless `script` is specified the inputs `structure`, "
+                "`potential` and  `parameters` have to be specified."
+            )
+
+        if "parameters" in self.ctx.inputs:
```

Comment:
The rest of this validation should really be done in an actual validator that is set on the relevant input port. The advantage is that the validation in this way is launched as soon as the process is submitted and if it is invalid, the nodes won't even be created. With the current approach, the process is always submitted and the node is created, and the process starts running, only to except in the first steps.

For example the first one, which is valdiating the `parameters` input, should really be done in the `LammpsCalculation` class.
```python
class LammpsBaseCalculation(CalcJob)

    @classmethod
    def define(cls, spec):
        super().define(spec)
        spec.input('parameter', validator=cls.validate_parameters)
    
    @clasmmethod
    def validate_parameters(cls, value, ctx) -> str | None:
        # validate here, value contains value passed to the port.
        # If there is a problem, return the error message as a string
```
The `ctx` variable will have the `Port` or `PortNamespace` that is being validated. This is useful if you need to validate a `PortNamespace`, for example the top-level inputs. You can use that to access multiple port values in case the validation depends on the value of multiple port.

Reply:
Sure thing will move those to the calculation, I was not aware of that difference, that is nice to know.